runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -367,7 +406,22 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) {
       LOG.info("Evaluating {}", transform);
       AppliedPTransform<?, ?, ?> appliedTransform = node.toAppliedPTransform();
       ctxt.setCurrentTransform(appliedTransform);
-      evaluator.evaluate(transform, ctxt);
+      // by default, we don't cache
+      boolean cacheHint = false;
+      // if the output of the node (aka PTransform) is already known in the cache candidates map
+      // (initialized with inputs), and it appears more than one time, then we enable caching
+
+      // considering the node output for caching
+      for (TaggedPValue output : node.getOutputs()) {
+        PValue value = output.getValue();
+        if ((value instanceof PCollection)
+            && cacheCandidates.containsKey(value)
+            && cacheCandidates.get(value) > 1) {
+          cacheHint = true;
+          break;
+        }
+      }
+      evaluator.evaluate(transform, ctxt, cacheHint);
```

Comment:
Mmm... and when a `Dataset` is checked-in, the `EvaluationContext` would check if its corresponding `PCollection` has a cache-hint and cache it ? sounds like it could work.
Though it is a bit nasty and relies on the whole static singleton in the Driver thing...
@jbonofre would you mind giving this a shot to see how this looks ? it could avoid tons of code that only passes around the hints.. which also has to be maintained etc.